Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,237 @@
+    : 0 ≤ deriv (-f) x := by
+  show 0 ≤ deriv (((fun x => -x) ∘ (fun x => f x))) x
+  rw [deriv.comp]
+  simp
+  apply h₀
+  tauto
+  refine Differentiable.differentiableAt ?hh₂.h
+  · exact differentiable_neg
+  · apply DifferentiableOn.differentiableAt
+    · exact hd₀
+    · refine Ioo_mem_nhds ?hh.hs.ha ?hh.hs.hb
+      · exact hx.1
+      · linarith[hx.2]
+
+/-- If `f'` is the derivative of `f` then  `0 ≤ f' x → (-f)' x ≤ 0`. -/
+theorem deriv_neg_nonpos {f : ℝ → ℝ} {b c : ℝ}
+  (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+  (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) (x : ℝ) :
+  x ∈ Set.Ioo b c → deriv (-f) x ≤ 0 := by
+        intro hx
+        show deriv (((fun x => -x) ∘ (fun x => f x))) x ≤ 0
+        rw [deriv.comp]
+        simp
+        apply h₁;tauto
+
+        apply differentiable_neg
+        apply DifferentiableOn.differentiableAt
+        exact hd₁
+        apply Ioo_mem_nhds
+        exact hx.1
+        linarith[hx.2]
+
+/-!
+### The First-Derivative Test
+
+Using the connection beetween monotonicity and derivatives we obtain the familiar
+First-Derivative Test from calculus.
+-/
+
+/-- If `f` is monotone on `(a,b]` and antitone on `[b,c)` then `f` has
+a local maximum at `b`. -/
+lemma isLocalMax_of_mono_anti.{u, v}
+  {α : Type u} [TopologicalSpace α] [LinearOrder α] [OrderClosedTopology α]
+    {β : Type v} [Preorder β]
+    {a b c : α} (g₀ : a < b) (g₁ : b < c)
+    {f : α → β}
+    (h₀ : MonotoneOn f (Set.Ioc a b))
+    (h₁ : AntitoneOn f (Set.Ico b c)) : IsLocalMax f b := by
+  unfold IsLocalMax IsMaxFilter Filter.Eventually
+  rw [nhds_def, Filter.mem_iInf]
+  exists {Set.Ioo a c}, (Set.toFinite _), (fun _ ↦ Set.Ioo a c ∪ {x | f x ≤ f b})
+  simp only [Set.mem_setOf_eq, Subtype.forall, Set.mem_singleton_iff, forall_eq, Set.mem_Ioo,
+    Set.iInter_coe_set, Set.iInter_iInter_eq_left]
+
+  constructor
+  apply Filter.mem_iInf_of_mem
+  · simp_all
+  · simp_all only [and_self, true_and]; apply isOpen_Ioo
+  · ext u
+    simp only [Set.mem_setOf_eq, Set.mem_union, Set.mem_Ioo, iff_or_self, and_imp]
+    intros
+    by_cases u < b
+    · apply h₀
+      · simp_all only [Set.mem_Ioc, true_and]
+        apply le_of_lt; tauto
+      · simp_all only [Set.mem_Ioc, le_refl, and_self]
+      · apply le_of_lt; tauto
+    · apply h₁
+      · simp_all
+      · simp_all
+      · apply le_of_not_lt
+        tauto
+
+ /-- The First-Derivative Test from calculus, maxima version.
+  Suppose `a < b < c`,
+    `f : ℝ → ℝ` is continuous at `b`,
+    the derivative `f'` is nonnegative on `(a,b)`, and
+    the derivative `f'` is nonpositive on `(b,c)`.
+  Then `f` has a local maximum at `a`. -/
+lemma first_derivative_test_max {f : ℝ → ℝ} {a b c : ℝ}
+    (g₀ : a < b) (g₁ : b < c)
+    (h : ContinuousAt f b)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b))
+    (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ :  ∀ x ∈ Set.Ioo a b, 0 ≤ deriv f x)
+    (h₁ :  ∀ x ∈ Set.Ioo b c, deriv f x ≤ 0)
+    : IsLocalMax f b := by
```

Comment:
Should we assume that `∀ᶠ x in 𝓝[≠] b, DifferentiableAt ℝ f x`, `∀ᶠ x in 𝓝[<] b, deriv f x ≤ 0`, and `∀ᶠ x in 𝓝[>] b, 0 ≤ deriv f x`.

Reply:
That makes sense. Maybe two versions, `first_derivative_test_max` and `first_derivative_test_max'` since mine is a bit easier to understand?

Reply:
Now I have added the precise statement you suggested, sorry for the delay.